### ISSUE TITLE
fix docker builds

### ISSUE THREAD

**Issue by casperdcl** [CCPPETMR/SIRF-SuperBuild#195]
- [ ] fix tests not running in docker
    + https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/483407710#L5255
    + [x] ensure future failures will not be silently ignored
- [x] ~~fix docker login~~
    + https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/483407710#L5234
    + ~~https://community.atlassian.com/t5/Bitbucket-questions/Piplenes-docker-login-can-not-perform-an-interactive-login-from/qaq-p/595736~~
    + https://discuss.bitrise.io/t/docker-login-failed-only-pr-trigger/2232
        * intentionally cannot build docker on PRs (security)

depends on #212 -> #213

**Comment by casperdcl** [comment]
ah this is actually caused by https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/483407710#L5089

@paskino what's the current quickest way to install the SuperBuild? I was using ubuntu16.04 with as many system libs as possible. Will I have to `-DUSE_SYSTEM_Boost=OFF` or can I get away with upgrading to 18.04 and using system libs?

**Comment by casperdcl** [comment]
sigh #91 would've fixed this issue

**Comment by KrisThielemans** [comment]
I'm fine with upgrading the docker build to 18.04. Otherwise loads of trouble with Gadgetron. In any case, it's a container, so why not?

**Comment by KrisThielemans** [comment]
@casperdcl, we're now ready for Ubuntu 18.04. Please use that (and the system boost and hdf5 etc). Currently, the docker build has a timeout.

**Comment by KrisThielemans** [comment]
> Currently, the docker build has a timeout.
Ignore this. I was checking the SIRF builds where the last job has a time-out. sorry

**Comment by casperdcl** [comment]
@KrisThielemans from `.travis.yml`, where do the following dependencies come from?

```yaml
- libopenblas-dev
- libatlas-base-dev
```

Are they required for gadgetron?

**Comment by KrisThielemans** [comment]
Not 100% sure. They're not mentioned on https://github.com/gadgetron/gadgetron/wiki/Linux-Installation. I think they are optional and will make some Gadgetron stuff faster. I see we have them in [the VM](https://github.com/CCPPETMR/CCPPETMR_VM/blob/bda900d9674ff8be05429cf7b41da8491f70c0eb/scripts/INSTALL_prerequisites_with_apt-get.sh#L22) as well (which I'm sure we copied into .travis.yml).

@rijobro has to use openblas on OSX as otherwise it found Apple's Accelerate library for BLAS which apparently had a bug. See https://github.com/gadgetron/gadgetron/issues/719

**Comment by casperdcl** [comment]
hmm they're not in docker atm, nor are -DUSE_ITK or ACE. Was trying to update things. Such #91 much fun.

**Comment by KrisThielemans** [comment]
ACE has to be there somehow. Otherwise it will fail AFAIK (could be by default I guess) .

ITK is optional but useful to read nifti's and Dicoms etc with STIR.

what about for docker cloning CCPPETMR_VM and use some of its scripts? I guess we could do the same for Travis but I suppose it'll be faster to list them explicitly (not sure).

**Comment by casperdcl** [comment]
I don't know - I'd be willing to deal with #91 now if you/@paskino are ready to merge the VM repo into this one, then it would be easier

**Comment by casperdcl** [comment]
argh I'm also running into https://github.com/conda/conda/issues/6030 now

**Comment by KrisThielemans** [comment]
re merging CCPPETMR_VM here. I guess we can do this while keeping the old repo, so I don't mind really. However moving everything to the proposed structure seems quite (too) ambitious... I'm afraid I cannot help really. no time. sorry.

**Comment by casperdcl** [comment]
OK I now have:

```
$ docker-compose -f docker-compose.yml -f docker-compose.nix.yml -f docker-compose.devel.yml build --build-arg UID=1000 --build-arg GROUPS=1000 sirf
Step 1/50 : FROM ubuntu:18.04 as base
...
Step 9/50 : FROM base as core
...
Step 31/50 : FROM core as sirf
...
Step 46/50 : RUN bash user_sirf-ubuntu.sh
$ git clone https://github.com/CCPPETMR/SIRF-SuperBuild --recursive -b master /opt/SIRF-SuperBuild
Cloning into '/opt/SIRF-SuperBuild'...
$ pushd /opt/SIRF-SuperBuild
/opt/SIRF-SuperBuild /home-away
$ echo $BUILD_FLAGS $EXTRA_BUILD_FLAGS
-DCMAKE_BUILD_TYPE=Release -DBUILD_STIR_WITH_OPENMP=ON -DUSE_SYSTEM_ACE=ON -DUSE_SYSTEM_Armadillo=ON -DUSE_SYSTEM_Boost=ON -DUSE_SYSTEM_FFTW3=ON -DUSE_SYSTEM_HDF5=ON -DUSE_ITK=ON -DBUILD_siemens_to_ismrmrd=ON -DUSE_SYSTEM_SWIG=ON -DDEVEL_BUILD=ON
$ cmake $BUILD_FLAGS $EXTRA_BUILD_FLAGS .
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /opt/pyvenv/bin/x86_64-conda_cos6-linux-gnu-cc
-- Check for working C compiler: /opt/pyvenv/bin/x86_64-conda_cos6-linux-gnu-cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/pyvenv/bin/x86_64-conda_cos6-linux-gnu-c++
-- Check for working CXX compiler: /opt/pyvenv/bin/x86_64-conda_cos6-linux-gnu-c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- EXTERNAL_PROJECT_BUILD_TYPE: Release
-- Found PythonInterp: /opt/pyvenv/bin/python (found version "2.7.15")
-- Found PYTHON_EXECUTABLE=/opt/pyvenv/bin/python
-- Python version 2.7.15
-- Found PythonLibs: /opt/pyvenv/lib/libpython2.7.so (found version "2.7.15")
-- Found PYTHON_INCLUDE_DIRS=/opt/pyvenv/include/python2.7
-- Found PYTHON_LIBRARIES=/opt/pyvenv/lib/libpython2.7.so
-- Could NOT find Matlab (missing:  Matlab_INCLUDE_DIRS Matlab_MEX_LIBRARY Matlab_MEX_EXTENSION Matlab_ROOT_DIR MAIN_PROGRAM)
-- Python libraries found
-- SIRF and/or STIR Python modules will be installed in /opt/SIRF-SuperBuild/INSTALL/python
-- SuperBuild - First pass
-- Matlab_ROOT_DIR=
-- STIR_DIR=
-- NiftyReg_Binary_DIR=
-- SuperBuild - First pass - done
-- SuperBuild - CCPPETMR => Requires SIRF, STIR, Gadgetron, siemens_to_ismrmrd, NiftyReg,
-- Matlab_ROOT_DIR=
-- STIR_DIR=
-- NiftyReg_Binary_DIR=
-- SuperBuild -   SIRF => Requires STIR, Boost, HDF5, ISMRMRD, FFTW3, SWIG, NiftyReg,
-- SuperBuild -     STIR => Requires Boost, ITK,
-- SuperBuild -       Boost[OK]
-- Boost version: 1.65.1
-- USING the system Boost, using BOOST_LIBRARY_DIR=
-- SuperBuild -       ITK => Requires HDF5,
-- SuperBuild -         HDF5[OK]
-- HDF5: Using hdf5 compiler wrapper to determine C configuration
-- Found HDF5: /usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5.so;pthread;sz;z;dl;m (found suitable version "1.10.0.1", minimum required is "1.8")
-- USING the system HDF5, found HDF5_INCLUDE_DIRS=/usr/include/hdf5/serial, HDF5_C_LIBRARY_hdf5=/usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5.so,HDF5_LIBRARIES=/usr/lib/x86_64-linux-gnu/hdf5/serial/libhdf5.so;pthread;sz;z;dl;m
-- SuperBuild -       ITK[OK]
-- Adding project ITK
-- SuperBuild -     STIR[OK]
-- Adding project STIR
-- SuperBuild -     ISMRMRD => Requires HDF5[INCLUDED], Boost[INCLUDED], FFTW3,
-- SuperBuild -       FFTW3[OK]
FFTW3 UNIX libraries: FFTW3_LIBRARY-NOTFOUND
CMake Error at /opt/cmake/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
  Could NOT find FFTW3 (missing: FFTW3_LIBRARY)
Call Stack (most recent call first):
  /opt/cmake/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  CMake/FindFFTW3.cmake:114 (find_package_handle_standard_args)
  SuperBuild/External_FFTW3.cmake:87 (find_package)
  CMake/ExternalProjectDependency.cmake:593 (include)
  SuperBuild/External_ISMRMRD.cmake:29 (ExternalProject_Include_Dependencies)
  CMake/ExternalProjectDependency.cmake:593 (include)
  SuperBuild/External_SIRF.cmake:37 (ExternalProject_Include_Dependencies)
  CMake/ExternalProjectDependency.cmake:593 (include)
  CMake/ExternalProjectDependency.cmake:639 (ExternalProject_Include_Dependencies)
  SuperBuild.cmake:210 (ExternalProject_Include_Dependencies)
  CMakeLists.txt:18 (include)

-- Configuring incomplete, errors occurred!
See also "/opt/SIRF-SuperBuild/CMakeFiles/CMakeOutput.log".
Makefile:45: recipe for target 'build' failed
```

Is this #147? I've already installed `libfftw3-dev`

**Comment by casperdcl** [comment]
nvm, seems to be caused by `conda install gxx_linux-64` which was my attempted fix of https://github.com/conda/conda/issues/6030

**Comment by casperdcl** [comment]
```
_reg_maths_eigen.cpp:(.text._Z3svdIfEvPPT_mmPS2_S3_S3_[_Z3svdIfEvPPT_mmPS2_S3_S3_]+0x5e9): undefined reference to `GOMP_parallel'
/opt/SIRF-SuperBuild/INSTALL/lib/lib_reg_maths.a(_reg_maths_eigen.cpp.o):_reg_maths_eigen.cpp:(.text._Z3svdIfEvPPT_mmPS2_S3_S3_[_Z3svdIfEvPPT_mmPS2_S3_S3_]+0x610): more und
efined references to `GOMP_parallel' follow
collect2: error: ld returned 1 exit status
src/Registration/cReg/CMakeFiles/sirf_tensor_split_join.dir/build.make:126: recipe for target 'src/Registration/cReg/sirf_tensor_split_join' failed
make[5]: *** [src/Registration/cReg/sirf_tensor_split_join] Error 1
CMakeFiles/Makefile2:758: recipe for target 'src/Registration/cReg/CMakeFiles/sirf_tensor_split_join.dir/all' failed
make[4]: *** [src/Registration/cReg/CMakeFiles/sirf_tensor_split_join.dir/all] Error 2
make[4]: *** Waiting for unfinished jobs....
[ 67%] Building CXX object src/xSTIR/pSTIR/CMakeFiles/_pystir.dir/opt/SIRF-SuperBuild/sources/STIR/src/modelling_buildblock/modelling_registries.cxx.o
[ 68%] Building CXX object src/xSTIR/pSTIR/CMakeFiles/_pystir.dir/opt/SIRF-SuperBuild/sources/STIR/src/spatial_transformation_buildblock/spatial_transformation_registries.c
xx.o
[ 69%] Linking CXX shared module _pystir.so
[ 69%] Built target _pystir
make[3]: *** [all] Error 2
Makefile:138: recipe for target 'all' failed
CMakeFiles/SIRF.dir/build.make:118: recipe for target 'builds/SIRF/stamp/SIRF-build' failed
make[2]: *** [builds/SIRF/stamp/SIRF-build] Error 2
make[1]: *** [CMakeFiles/SIRF.dir/all] Error 2
CMakeFiles/Makefile2:306: recipe for target 'CMakeFiles/SIRF.dir/all' failed
Makefile:94: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by KrisThielemans** [comment]
hmmm. I suspect this is a NiftyReg issue. Maybe someone with Ubuntu 18.04 (non-docker) could confirm this.

Their [CMake file](https://github.com/KCL-BMEIS/niftyreg/blob/731a565bd42ca97ff5968adb1c06133ea72f0856/CMakeLists.txt#L173 doesn't add linker flags).  I had the same in STIR where the `-fopenmp` flag wasn't passed, and added [a work-around](https://github.com/UCL/STIR/blob/63c8f6ffdc3ea7674fa5223c8cc45a37145dc3f9/src/CMakeLists.txt#L147), but later found out this was the wrong thing to do: https://github.com/UCL/STIR/issues/292.

Suggested work-around for now: add `-DUSE_OPENMP:BOOL=OFF` to NiftyReg build. If that works, could you then create a separate issue to resolve this later?

**Comment by casperdcl** [comment]
Hah that solved it.
